### PR TITLE
circuits: zk-gadgets: arithmetic, fixed-point, wallet-operations: Refactor and add unit tests

### PR DESCRIPTION
**Pull request by joeykraut**
### Purpose
This PR refactors the `arithmetic`, `fixed-point`, and `wallet-operations` modules of the `zk-gadgets` to work over the new `mpc-jellyfish` system. I also added unit tests for all refactored gadgets, some are tested implicitly.

### Testing
- Unit tests pass

### CODE REVIEW

File: circuits/src/zk_gadgets/fixed_point.rs [hunk 1]
```diff
@@ -101,46 +101,147 @@ pub struct MultiproverFixedPointGadget;
...
-            diff,
-            fabric,
-            cs,
+        let diff = cs.lc(
+            &[one_var, lhs_minus_rhs, zero_var, zero_var],
+            &[*TWO_TO_M_SCALAR, -one, one, one],
+        )?;
+
+        MultiproverGreaterThanEqZeroGadget::<DEFAULT_FP_PRECISION>::constrain_greater_than_zero(
+            diff, fabric, cs,
```
Comment (akirillo) [inline circuits/src/zk_gadgets/fixed_point.rs:131]: this seems to be a different constraint than what was there previously - seems like it used to be `2^m - 1 > (lhs - rhs*2^m)`, and is now `2^m - (lhs - rhs*2^m) > 0`. am i reading this wrong?

**Review by akirillo — APPROVED** [review]
some small concerns but nothing that requires further discussion so approving to unblock

File: circuits/src/zk_gadgets/arithmetic.rs [hunk 2]
```diff
@@ -29,347 +27,122 @@ use super::select::CondSelectGadget;
...
-            let (_, _, out_var1) = cs.multiply(recursive_result.clone(), recursive_result);
-            let (_, _, out_var2) = cs.multiply(out_var1.into(), x_lc);
-            out_var2.into()
-        }
-    }
-}
-
-/// An exponentiation gadget on a private exponent; compute x^\alpha
-#[derive(Clone, Debug)]
-pub struct PrivateExpGadget<const ALPHA_BITS: usize> {}
```
[the remark this answers is not in the document]
Comment (joeykraut) [reply circuits/src/zk_gadgets/arithmetic.rs:99]: correct, it was for elgamal which is gone

File: circuits/src/zk_gadgets/arithmetic.rs [hunk 3]
```diff
@@ -29,347 +27,122 @@ use super::select::CondSelectGadget;
...
+    ) -> Result<(Variable, Variable), CircuitError> {
+        let a_bigint = scalar_to_biguint(&a.eval(cs));
+        let b_bigint = scalar_to_biguint(&b.eval(cs));
 
         // Compute the divrem outside of the circuit
-        // Verifier evals all wires to zero -- it only knows commitments
-        // Handle this case explicitly
         let (q, r) = if b_bigint == BigUint::zero() {
-            (BigUint::zero(), b_bigint)
+            (BigUint::zero(), a_bigint)
```
[the remark this answers is not in the document]
Comment (joeykraut) [reply circuits/src/zk_gadgets/arithmetic.rs:40]: The operation is perfectly well defined when `b = 0`, this was a bug before. To be clear, setting `r = 0` does not _make_ the constraints fail. Remember that the prover may assign these wires however they like, it is simply that we assign them honestly in our codebase. The constraints will be able to be satisfied when `b = 0` and that is okay from our perspective

File: circuits/src/zk_gadgets/fixed_point.rs (hunk 1 above)
Comment (joeykraut) [reply circuits/src/zk_gadgets/fixed_point.rs:131]: It was `2^m - 1 >= (lhs - rhs << m)`, rearranging the terms gives:
`2^m - (lhs - rhs*2^m) - 1 >= 0`. Ours is a discrete domain so this is equivalent to the constraint:
`2^m - (lhs - rhs * 2^m) > 0` which is what we have here

Comment (akirillo) [reply circuits/src/zk_gadgets/fixed_point.rs:131]: ah, i missed the `>=`, thank you

**Pull request merged**